File: Data Structures & Algorithms/max-area-of-island/submission-2.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        max_area = 0
        ROWS, COLS = len(grid), len(grid[0])

        def dfs(r, c, area):
            nonlocal max_area

            if grid[r][c] == 0:
                return

            area[0] += 1
            max_area = max(max_area, area[0])
            grid[r][c] = 0
            if r + 1 < ROWS: dfs(r + 1, c, area)
            if r - 1 >= 0: dfs(r - 1, c, area)
            if c + 1 < COLS: dfs(r, c + 1, area)
            if c - 1 >= 0: dfs(r, c - 1, area)

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1:
                    dfs(r, c, [0])
        
        return max_area

File: Data Structures & Algorithms/count-number-of-islands/submission-2.java
class Solution {
    private int islands = 0;

    public int numIslands(char[][] grid) {
        int ROWS = grid.length, COLS = grid[0].length;

        for (int r = 0; r < ROWS; r++)
        {
            for (int c = 0; c < COLS; c++)
            {
                if (grid[r][c] == '1')
                {
                    dfs(r, c, grid);
                    islands++;
                }
            }
        }

        return islands;
    }

    private void dfs(int r, int c, char[][] grid)
    {
        int ROWS = grid.length, COLS = grid[0].length;

        if (grid[r][c] == '0')
        {
            return;
        }

        grid[r][c] = '0';
        if (r + 1 < ROWS) dfs(r + 1, c, grid);
        if (r - 1 >= 0) dfs(r - 1, c, grid);
        if (c + 1 < COLS) dfs(r, c + 1, grid);
        if (c - 1 >= 0) dfs(r, c - 1, grid);
    }
}
